File: php/3-longest-substring-without-repeating-characters.php
<?php
class Solution 
{
    private $freq = [];
    /**
     * @param String $s
     * @return Integer
     */
    public function lengthOfLongestSubstring($s) 
    {
        $i = 0;
        $j = -1;
        $len = strlen($s);
        $maxLen = 0;
        while ($i < $len) {
            if ($j+1 < $len && $this->getFreq($s[$j+1]) == 0) {
                $this->freq[$s[++$j]]++;
            } else {
                $this->freq[$s[$i++]]--;
            }
            $maxLen = max($maxLen, $j-$i+1);
        }
        return $maxLen;
    }
    
    private function getFreq($s)
    {
        if (!isset($this->freq[$s])) {
            $this->freq[$s] = 0;
        }
        return $this->freq[$s];
    }
}

//test
$str  = "abcabcbb";
$s = new Solution();
echo $s->lengthOfLongestSubstring($str);

File: go/3-longest-substring-without-repeating-characters.go
package main

func lengthOfLongestSubstring(s string) int {
	record := make(map[string]int, len(s))
	maxLen := 0
	i := 0
	j := 0
	for j < len(s) {
		if _, exist := record[s[j:j+1]]; !exist {
			record[s[j:j+1]]++
			j++
		} else {
			delete(record, s[i:i+1])
			i++
		}
		maxLen = int(math.Max(float64(maxLen), float64(j-i)))
	}
	return maxLen
}
